Report running DDLWorker hosts

tuanpach opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Report running DDLWorker hosts by creating replica_dir and mark replicas active in DDLWorker

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

tuanpach committed:
1) Create replica_dir in DDLWorker
2) Mark replicas active in DDLWorker

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #69658 from tuanpach/ClickHouse:report-running-ddl-worker-hosts -->
*This is an automated comment for commit 12c833ff17cf1059d4d30c2a648240b5849cd50a with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/69658/12c833ff17cf1059d4d30c2a648240b5849cd50a/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/7b94dc1813ae2936fd0e276eb46b45ef6fcae994/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

tuanpach committed:
Add test_ddl_worker_replicas

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Add distributed_ddl.replicas_path setting

tavplubix commented on src/Interpreters/DDLWorker.cpp:
```diff
@@ -1222,6 +1234,97 @@ void DDLWorker::runMainThread()
+
+        try
+        {
+            HostID host = HostID::fromString(host_id);
+            /// The port is considered local if it matches TCP or TCP secure port that the server is listening.
+            bool is_local_host = (maybe_secure_port && host.isLocalAddress(*maybe_secure_port)) || host.isLocalAddress(port);
```
See also `DDLTask::findCurrentHostID`

tavplubix commented on src/Interpreters/DDLWorker.h:
```diff
@@ -201,6 +214,8 @@ class DDLWorker
     const CurrentMetrics::Metric * max_entry_metric;
     const CurrentMetrics::Metric * max_pushed_entry_metric;
+
+    std::unordered_map<String, std::pair<ZooKeeperPtr, zkutil::EphemeralNodeHolderPtr>> active_node_holders;
```
Maybe we can use `host_fqdn_id` and create one active node per replica, but I'm not sure

tuanpach replied:
Multiple host_id can point to the same instance. For example, both `domain1-host2.aws.com` and `domain2-host2.aws.com` are resolved to `host2`. 

So my idea is that for each host_id in each cluster, the DDL worker creates a `host_id` dir in `replicas_dir` in `initializeReplication`. In the main thread, each DDL worker checks if the `host_id` under `replicas_dir` belongs to itself, it will create an `active` node under the `host_id` dir to indicate that the `host_id` is running.

tuanpach replied:
Because multiple host_ids can belong to 1 instance, I use a map to hold the ephemeral nodes.

tavplubix replied:
Ok, it makes sense. Without creating multiple active nodes we would have to perform complex checks (like in `findCurrentHostID`) to check if a host is active or not

tuanpach commented on src/Interpreters/DDLWorker.cpp:
```diff
@@ -1222,6 +1234,97 @@ void DDLWorker::runMainThread()
+        }
+
+        /// Create "active" node (remove previous one if necessary)
+        String active_path = fs::path(replicas_dir) / host_id / "active";
```
I also prefer to have a dir for a host. The logic to check offline hosts is the same as ReplicatedDatabase, and we can add more attributes to the host in the future.

tuanpach commented on tests/integration/test_ddl_worker_replicas/test.py:
```diff
@@ -0,0 +1,67 @@
+    assert expected_replicas.sort() == replica_list.sort()
+
+    for replica in replica_list:
+        result = node1.query(
+            f"SELECT name, value, ephemeralOwner FROM system.zookeeper WHERE path='/clickhouse/task_queue/replicas/{replica}'"
```
When I ran two server instances locally, in `system.clusters`, I could only find `ReplicatedDatabase` clusters, not DDL clusters.
After I dropped all replicated databases, the following query showed an empty result

```
 SELECT host_name, port, cluster FROM system.clusters WHERE is_active==True
 ```
 
 Is there anything that I missed?

tavplubix replied:
Those clusters have to be declared in config.xml: https://clickhouse.com/docs/en/operations/server-configuration-parameters/settings#server-settings-remote-servers
For example: https://github.com/ClickHouse/ClickHouse/blob/f49412e967ec549d8a49e53ce3bfba27e13994fc/tests/config/config.d/clusters.xml

tuanpach committed:
Remove sleep in test_ddl_worker_replicas test

tuanpach committed:
Reformat the test

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Not re-create an ephemeral node if it exists

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Fix style check error

tuanpach committed:
1) Extend stop_start_wait_sec and reduce connection_timeout_ms in
test_replicated_database when restarting an instance.
2) Make createReplicaDirs and markReplicasActive virtual functions

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts

tuanpach committed:
Fix flaky test

tuanpach committed:
Revert changes in test_replicated_database

tuanpach committed:
Make test_ddl_worker_replicas predictable

tuanpach committed:
Merge branch 'master' into report-running-ddl-worker-hosts
